File: src/main/java/org/jukeboxmc/command/CommandManager.java
package org.jukeboxmc.command;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jukeboxmc.command.internal.*;

import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class CommandManager {

    private final List<Command> commands = new ArrayList<>();

    public CommandManager() {
        List<Class<? extends Command>> commands = Arrays.asList(
                PluginsCommand.class,
                GameModeCommand.class,
                OperatorCommand.class,
                RemoveOperatorCommand.class,
                TeleportCommand.class,
                StopCommand.class,
                SaveCommand.class,
                KillCommand.class,
                KickCommand.class,
                EffectCommand.class,
                EnchantCommand.class,
                GameRuleCommand.class,
                StatusCommand.class,
                TimeCommand.class
        );

        for( Class<? extends Command> commandClass : commands ) {
            try {
                this.registerCommand( commandClass.getConstructor().newInstance() );
            } catch ( InstantiationException | IllegalAccessException | NoSuchMethodException |
                      InvocationTargetException e ) {
                e.printStackTrace();
            }
        }
    }

    public void handleCommandInput(@NotNull CommandSender commandSender, @NotNull String input ) {
        try {
            final String[] commandParts = input.substring( 1 ).split( " " );
            final String commandIdentifier = commandParts[0];

            int consumed = 0;
            Command targetCommand = null;
            while ( targetCommand == null ) {
                for ( Command command : this.commands ) {
                    if ( command.getCommandData().getName().equalsIgnoreCase( commandIdentifier ) || ( command.getCommandData().getAliases() != null && command.getCommandData().getAliases().contains( commandIdentifier ) ) ) {
                        targetCommand = command;
                        break;
                    }
                }

                consumed++;
                if ( targetCommand == null ) {
                    if ( commandParts.length == consumed ) {
                        break;
                    }
                }
            }

            if ( targetCommand == null ) {
                commandSender.sendMessage( "The command for " + commandIdentifier + " could not be found" );
                return;
            }

            if ( targetCommand.getCommandData().getPermission() != null && !commandSender.hasPermission( targetCommand.getCommandData().getPermission() ) ) {
                if ( targetCommand.getCommandData().getPermissionMessage() != null && !targetCommand.getCommandData().getPermissionMessage().isEmpty() ) {
                    commandSender.sendMessage( targetCommand.getCommandData().getPermissionMessage() );
                } else {
                    commandSender.sendMessage( "§cYou don't have permission to do that" );
                }
                return;
            }

            String[] params;
            if ( commandParts.length > consumed ) {
                params = new String[commandParts.length - consumed];
                System.arraycopy( commandParts, consumed, params, 0, commandParts.length - consumed );
            } else {
                params = new String[0];
            }
            targetCommand.execute( commandSender, commandIdentifier, params );
        } catch ( Throwable e ) {
            e.printStackTrace();
        }
    }

    public void registerCommand( Command command ) {
        this.commands.add( command );
    }

    public void unregisterCommand( Command command ) {
        this.commands.remove( command );
    }

    public boolean isCommandRegistered( String name ) {
        for ( Command command : this.commands ) {
            if ( command.getCommandData().getName().equalsIgnoreCase( name ) ) {
                return true;
            }
        }

        return false;
    }

    public @NotNull List<Command> getCommands() {
        return new ArrayList<>( this.commands );
    }

    private @Nullable Command getCommandByNameOrAlias(String identifier ) {
        for( Command command : this.commands ) {
            if ( command.getCommandData().getName().equalsIgnoreCase( identifier ) || ( command.getCommandData().getAliases() != null &&
                    command.getCommandData().getAliases().contains( identifier ) ) ) {
                return command;
            }
        }

        return null;
    }
}

File: src/main/java/org/jukeboxmc/command/CommandManager.kt
package org.jukeboxmc.command

import org.jukeboxmc.command.internal.EffectCommand
import org.jukeboxmc.command.internal.EnchantCommand
import org.jukeboxmc.command.internal.GameModeCommand
import org.jukeboxmc.command.internal.GameRuleCommand
import org.jukeboxmc.command.internal.KickCommand
import org.jukeboxmc.command.internal.KillCommand
import org.jukeboxmc.command.internal.OperatorCommand
import org.jukeboxmc.command.internal.PluginsCommand
import org.jukeboxmc.command.internal.RemoveOperatorCommand
import org.jukeboxmc.command.internal.SaveCommand
import org.jukeboxmc.command.internal.StatusCommand
import org.jukeboxmc.command.internal.StopCommand
import org.jukeboxmc.command.internal.TeleportCommand
import org.jukeboxmc.command.internal.TimeCommand

/**
 * @author LucGamesYT
 * @version 1.0
 */
class CommandManager {
    private val commands: MutableList<Command> = ArrayList()

    init {
        val commands = listOf(
            PluginsCommand::class.java,
            GameModeCommand::class.java,
            OperatorCommand::class.java,
            RemoveOperatorCommand::class.java,
            TeleportCommand::class.java,
            StopCommand::class.java,
            SaveCommand::class.java,
            KillCommand::class.java,
            KickCommand::class.java,
            EffectCommand::class.java,
            EnchantCommand::class.java,
            GameRuleCommand::class.java,
            StatusCommand::class.java,
            TimeCommand::class.java,
        )
        commands.forEach { commandClass ->
            registerCommand(commandClass.getConstructor().newInstance())
        }
    }

    fun handleCommandInput(commandSender: CommandSender, input: String) {
        val commandParts = input.substring(1).split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val commandIdentifier = commandParts[0]
        var consumed = 0
        var targetCommand: Command? = null
        while (targetCommand == null) {
            for (command in commands) {
                if (command.commandData.name.equals(
                        commandIdentifier,
                        ignoreCase = true,
                    ) || command.commandData.getAliases().contains(
                        commandIdentifier,
                    )
                ) {
                    targetCommand = command
                    break
                }
            }
            consumed++
            if (targetCommand == null) {
                if (commandParts.size == consumed) {
                    break
                }
            }
        }
        if (targetCommand == null) {
            commandSender.sendMessage("The command for $commandIdentifier could not be found")
            return
        }
        if (targetCommand.commandData.getPermission() != null && !commandSender.hasPermission(targetCommand.commandData.getPermission())) {
            if (targetCommand.commandData.getPermissionMessage().isNotEmpty()) {
                commandSender.sendMessage(targetCommand.commandData.getPermissionMessage())
            } else {
                commandSender.sendMessage("§cYou don't have permission to do that")
            }
            return
        }
        val params: Array<String>
        if (commandParts.size > consumed) {
            params = Array(commandParts.size - consumed) { "" }
            System.arraycopy(commandParts, consumed, params, 0, commandParts.size - consumed)
        } else {
            params = emptyArray()
        }
        targetCommand.execute(commandSender, commandIdentifier, params)
    }

    fun registerCommand(command: Command) {
        commands.add(command)
    }

    fun unregisterCommand(command: Command) {
        commands.remove(command)
    }

    fun isCommandRegistered(name: String?): Boolean {
        for (command in commands) {
            if (command.commandData.name.equals(name, ignoreCase = true)) {
                return true
            }
        }
        return false
    }

    fun getCommands(): List<Command> {
        return ArrayList(commands)
    }

    private fun getCommandByNameOrAlias(identifier: String): Command? {
        for (command in commands) {
            if (command.commandData.name.equals(
                    identifier,
                    ignoreCase = true,
                ) || command.commandData.getAliases().contains(identifier)
            ) {
                return command
            }
        }
        return null
    }
}
